### PR TITLE
Fix passing negative value as count to function returning counted_by

### LINKED ISSUE #13550: [BoundsSafety] passing negative value as count to function returning `counted_by` does not error
Setting the length of a count to a known negative value is an error, but the error doesn't currently cover the case when the pointer is the return value of a function. Example (https://godbolt.org/z/15sfWEdhY):
```
int * __counted_by(len) foo(int len);
void bar(int * __counted_by(len) p, int len);

void baz(int * __counted_by(*len) *p, int *len) {
    *p = (void*)0; *len = -1;                              // error
}

void test() {
    int *p1 = foo(-1);                                     // no error!
    bar(p1, -1);                                           // error
    int len = -1; int * __counted_by(len) p2 = (void*)0;   // error
}
```

Related: rdar://80808704

### PR DESCRIPTION
**Pull request by pandaaured**
Added check for value of return type's count expression such that no negative values are passed as arguments to
functions returning `counted_by`.

Changed tests to correspond with correct program behavior and removed rdar://80808704 and rdar://103368466 comments from the tests which errored.

Fixes: #13550

### CODE REVIEW

File: clang/lib/Sema/SemaExpr.cpp [hunk 1]
```diff
@@ -26523,6 +26511,26 @@ ExprResult Sema::ActOnBoundsSafetyCall(ExprResult Call) {
     ReturnCount = DefaultLvalueConversion(ReturnCount.get());
     if (!ReturnCount.get())
       return ExprError();
+
+    Expr *CountE = ReturnCount.get();
+    if (auto *OVE = dyn_cast<OpaqueValueExpr>(CountE))
+      CountE = OVE->getSourceExpr();
+    CountE = CountE->IgnoreParenCasts();
```
Comment (hnrklssn) [inline clang/lib/Sema/SemaExpr.cpp:26518]: Do paren casts prevent const eval? If so, this should probably be `IgnoreParenImpCasts`. Otherwise we should be able to skip it

Comment (ojhunt) [reply clang/lib/Sema/SemaExpr.cpp:26518]: Would be worth trying to come up with a test that would show the behavioral change

File: clang/test/BoundsSafety/Sema/arithmetic-ops-in-counted-by-calls.c [hunk 2]
```diff
@@ -71,6 +71,6 @@ void call_param_with_shared_size(void) {
 void *__sized_by(count * size) return_with_count_size(int count, int size);
 
 void call_return_with_count_size(void) {
-  // FIXME: rdar://103368466
+  // expected-error@+1{{negative size value of -1 for 'void *__single __sized_by(count * size)' (aka 'void *__single')}}
```
Comment (hnrklssn) [inline clang/test/BoundsSafety/Sema/arithmetic-ops-in-counted-by-calls.c:74]: I suggest we add another option to the select in `err_bounds_safety_dynamic_count_negative` so that we can emit `for return value of type ...`

File: clang/test/BoundsSafety-legacy-checks/CodeGen/counted_by_to_counted_by-O2.c [hunk 3]
```diff
@@ -71,7 +71,8 @@ void TestUCountNegFail(void) {
 // CHECK-NEXT:    unreachable, {{!annotation ![0-9]+}}
 //
 void TestCountNegRetFail(void) {
-  int *local_p = my_alloc_int(-1); //rdar://80808704
+  int x = -1;
+  int *local_p = my_alloc_int(x);
```
Comment (ojhunt) [inline clang/test/BoundsSafety-legacy-checks/CodeGen/counted_by_to_counted_by-O2.c:75]: This removed the 80808704 rdar reference, but I can't see a similar reference in the Sema tests so I'm not sure which of the tests this matches.

Part of me actually just wants the original test case copied verbatim into the Sema tests to make it unambiguous that it fails in Sema. (this is not a strongly held opinion)

@hnrklssn thoughts?

**Review by ojhunt — COMMENTED** [review]
I don't have code change feedback, I'm assuming the new error messages in the Sema tests correspond to the required codegen changes, but want to be sure that that is the case.

File: clang/test/BoundsSafety-legacy-checks/CodeGen/counted_by_to_counted_by-O2.c (hunk 3 above)
Comment (hnrklssn) [reply clang/test/BoundsSafety-legacy-checks/CodeGen/counted_by_to_counted_by-O2.c:75]: As in the entire file?

File: clang/lib/Sema/SemaExpr.cpp (hunk 1 above)
Comment (pandaaured) [reply clang/lib/Sema/SemaExpr.cpp:26518]: I wrote a few tests for the function with IgnoreParenCasts dropped that test various arithmetic expressions for negative counts, one including additional parentheses to check whether the evaluation is affected by this.
```C
int *__counted_by(len * count) foo(int len, short count);
int *__counted_by(len * count) baz(unsigned int len, long long count);

int *__counted_by((int)(len * count)) mul(unsigned len, unsigned count);
int *__counted_by((int)((len) * (count))) mul_paren(unsigned len, unsigned count);

void Test(void) {
  // expected-error@+1{{negative count value of -1 for 'int *__single __counted_by(len * count)'}}
  foo(1, -1);

  // expected-error@+1{{negative count value of -4294967295 for 'int *__single __counted_by(len * count)'}}
  baz(UINT_MAX, -1LL);

  // expected-error@+1{{negative count value of -1 for 'int *__single __counted_by((int)(len * count))'}}
  mul(0xFFFFFFFFu, 1u);

  // expected-error@+1{{negative count value of -1 for 'int *__single __counted_by((int)(len * count))' (aka 'int *__single')}}
  mul(0xFFFFFFFFu, 1u);
}
```
They should all error. I haven't committed them because I want to be sure that the tests are of the right scope and sufficiently broad for the desired use before I commit anything. Building clang and running clang-test with the above yielded no unexpectedly failing tests that were related to the change in Sema.

**Comment by pandaaured** [comment]
Removed the paren cast since I could not find any test cases where it made a difference in the evaluation. Let me know if I should add more test cases to check for different edge cases for overflowing arithmetic expressions that evaluate to constant time negative expressions; I only wrote a few and am unsure if there could be more thorough coverage there.

Have not yet implemented Henrik's suggestion to modify the diagnostics system since I am not very familiar with it yet. Would such a change be as simple as updating the `DiagnosticSemaKinds.td` file?
